--- a/scheduler/main.c	2019-12-13 15:44:10.000000000 +0100
+++ b/scheduler/main.c	2019-12-31 16:01:57.193166435 +0100
@@ -681,8 +681,15 @@
 
 #ifdef HAVE_ONDEMAND
   if (OnDemand)
+  {
     cupsdAddEvent(CUPSD_EVENT_SERVER_STARTED, NULL, NULL, "Scheduler started on demand.");
-  else
+# ifdef HAVE_SYSTEMD
+    sd_notifyf(0, "READY=1\n"
+               "STATUS=Scheduler is running...\n"
+               "MAINPID=%lu",
+               (unsigned long) getpid());
+# endif /* HAVE_SYSTEMD */
+  } else
 #endif /* HAVE_ONDEMAND */
   if (fg)
     cupsdAddEvent(CUPSD_EVENT_SERVER_STARTED, NULL, NULL, "Scheduler started in foreground.");
--- a/scheduler/org.cups.cupsd.service.in	2019-12-13 15:44:10.000000000 +0100
+++ b/scheduler/org.cups.cupsd.service.in	2019-12-31 16:01:57.193166435 +0100
@@ -1,11 +1,11 @@
 [Unit]
 Description=CUPS Scheduler
 Documentation=man:cupsd(8)
-After=sssd.service
+After=sssd.service network.target
 
 [Service]
 ExecStart=@sbindir@/cupsd -l
-Type=simple
+Type=notify
 Restart=on-failure
 
 [Install]
