--- a/configure.ac	2017-09-09 18:07:33.000000000 +0200
+++ b/configure.ac	2017-09-21 17:26:55.027360288 +0200
@@ -98,36 +98,6 @@
 AC_SUBST(GMODULE_CFLAGS)
 AC_SUBST(GMODULE_LIBS)
 
-PKG_CHECK_MODULES(LIBGVC, libgvc >= $LIBGVC_REQUIRED)
-AC_MSG_CHECKING([for CGRAPH])
-cgraph_tmp_LIBADD="$LIBADD"
-cgraph_tmp_CFLAGS="$CFLAGS"
-LIBADD="$LIBADD $LIBGVC_LIBS"
-CFLAGS="$CFLAGS $LIBGVC_CFLAGS"
-AC_RUN_IFELSE(
-	[AC_LANG_SOURCE([
-		#include <gvc.h>
-
-		int main(void) {
-			#ifdef WITH_CGRAPH
-				return 0;
-			#else
-				return -1;
-			#endif
-		}
-	])], [
-		AC_MSG_RESULT([yes])
-		VALAFLAGS="$VALAFLAGS -D WITH_CGRAPH"
-		have_cgraph=yes
-	], [
-		AC_MSG_RESULT([no])
-		have_cgraph=no
-	]
-)
-LIBADD="$cgraph_tmp_LIBADD"
-CFLAGS="$cgraph_tmp_CFLAGS"
-AM_CONDITIONAL(HAVE_CGRAPH, test "$have_cgraph" = "yes")
-
 AC_PATH_PROG([XSLTPROC], [xsltproc], :)
 AM_CONDITIONAL(HAVE_XSLTPROC, test "$XSLTPROC" != :)
 
@@ -156,16 +126,6 @@
            vapigen/Makefile
            vapigen/vala-gen-introspect/Makefile
            vapigen/vala-gen-introspect/vala-gen-introspect
-           libvaladoc/valadoc.pc
-           libvaladoc/valadoc.deps
-           libvaladoc/Makefile
-           valadoc/Makefile
-           valadoc/doclets/Makefile
-           valadoc/doclets/html/Makefile
-           valadoc/doclets/devhelp/Makefile
-           valadoc/doclets/gtkdoc/Makefile
-           valadoc/icons/Makefile
-           valadoc/tests/Makefile
 ])
 
 AC_OUTPUT
--- a/doc/Makefile.am	2017-09-09 18:07:33.000000000 +0200
+++ b/doc/Makefile.am	2017-09-21 17:27:06.763468209 +0200
@@ -70,10 +70,6 @@
 	--basedir=internal-apis/ \
 	$(NULL)
 
-if HAVE_CGRAPH
-COMMON_VALADOCFLAGS += -D WITH_CGRAPH
-endif
-
 gee_VALASOURCES = $(top_srcdir)/gee/*.vala
 internal-apis/gee: $(gee_VALASOURCES)
 	$(AM_V_GEN)$(top_builddir)/valadoc/valadoc \
--- a/Makefile.am	2017-09-09 18:07:33.000000000 +0200
+++ b/Makefile.am	2017-09-21 17:26:55.028360297 +0200
@@ -13,8 +13,6 @@
 	doc \
 	gobject-introspection \
 	vapigen \
-	libvaladoc \
-	valadoc \
 	$(NULL)
 
 if ENABLE_UNVERSIONED
