--- a/lib/Driver/ToolChains/Linux.cpp	2019-12-11 20:15:30.000000000 +0100
+++ b/lib/Driver/ToolChains/Linux.cpp	2019-12-29 18:22:06.552096476 +0100
@@ -205,7 +205,7 @@
   if (Triple.getArch() == llvm::Triple::riscv32)
     return "lib32";
 
-  return Triple.isArch32Bit() ? "lib" : "lib64";
+  return Triple.isArch32Bit() ? "lib32" : "lib";
 }
 
 static void addMultilibsFilePaths(const Driver &D, const MultilibSet &Multilibs,
@@ -317,8 +317,15 @@
   const std::string OSLibDir = getOSLibDir(Triple, Args);
   const std::string MultiarchTriple = getMultiarchTriple(D, Triple, SysRoot);
 
+  SmallString<128> libSuffix((Arch == llvm::Triple::x86) ? "i386" : "x86_64");
+  SmallString<128> crtSysPath(D.ResourceDir);
+
+  llvm::sys::path::append(crtSysPath, libSuffix);
+
+  addPathIfExists(D, crtSysPath, Paths);
+
   // Add the multilib suffixed paths where they are available.
-  if (GCCInstallation.isValid()) {
+  if (false && GCCInstallation.isValid()) {
     const llvm::Triple &GCCTriple = GCCInstallation.getTriple();
     const std::string &LibPath = GCCInstallation.getParentLibPath();
 
@@ -628,13 +635,13 @@
     Loader = "ld64.so.1";
     break;
   case llvm::Triple::x86:
-    LibDir = "lib";
+    LibDir = "lib32";
     Loader = "ld-linux.so.2";
     break;
   case llvm::Triple::x86_64: {
     bool X32 = Triple.getEnvironment() == llvm::Triple::GNUX32;
 
-    LibDir = X32 ? "libx32" : "lib64";
+    LibDir = X32 ? "libx32" : "lib";
     Loader = X32 ? "ld-linux-x32.so.2" : "ld-linux-x86-64.so.2";
     break;
   }
