diff -Naur a/src/core/api/qtbug-60565.cpp b/src/core/api/qtbug-60565.cpp
--- a/src/core/api/qtbug-60565.cpp	2017-05-19 06:22:04.000000000 +0200
+++ b/src/core/api/qtbug-60565.cpp	2017-05-31 18:13:18.833427261 +0200
@@ -87,21 +87,22 @@
 static void __shimCppDelete(void *address);
 static void __shimCppDeleteArray(void *address);
 
-static void* ShimCppNew(size_t size) {
+void* __attribute__ ((visibility ("hidden"))) ShimCppNew(size_t size) {
     return __shimCppNew(size);
 }
 
-static void* ShimCppNewArray(size_t size) {
+void* __attribute__ ((visibility ("hidden"))) ShimCppNewArray(size_t size) {
     return __shimCppNewArray(size);
 }
 
-static void ShimCppDelete(void* address) {
+void __attribute__ ((visibility ("hidden"))) ShimCppDelete(void* address) {
     __shimCppDelete(address);
 }
 
-static void ShimCppDeleteArray(void* address) {
+void __attribute__ ((visibility ("hidden"))) ShimCppDeleteArray(void* address) {
     __shimCppDeleteArray(address);
 }
+
 } // extern "C"
 
 static void* __shimCppNew(size_t size) {
