--- a/prnt/hpps/hppsfilter.c	2020-11-30 01:03:52.000000000 +0100
+++ b/prnt/hpps/hppsfilter.c	2021-01-02 20:42:52.275016577 +0100
@@ -104,7 +104,7 @@
     if(ptempbooklet_file == NULL)
     {
             fprintf(stderr, "ERROR: Unable to open temp file %s\n", temp_filename);
-            return 1;
+            return;
     }  
     chmod(temp_filename, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
 
--- a/scan/sane/hpaio.c	2020-11-30 01:03:52.000000000 +0100
+++ b/scan/sane/hpaio.c	2021-01-02 20:43:20.535080068 +0100
@@ -51,7 +51,7 @@
 #include "escl.h"
 #include "io.h"
 #include "orblitei.h"
-
+#include "orblite.h"
 
 #define DEBUG_DECLARE_ONLY
 #include "sanei_debug.h"
